### PR TITLE
Add Advanced render box styles

### PR DESCRIPTION
**Pull request by angerman**
Adds support for `--build-info` flags as proposed in #134. Also adds our all time favorite unicorn.

### CODE REVIEW

**Comment by angerman** [comment]
@ndmitchell This doesn't work all the time, as I'm running into:
```
Error when running Shake build system:
* inplace/bin/genprimopcode
* OracleQ (PackageDataKey (".build/stage0/utils/genprimopcode/package-data.mk",".build_stage0_utils_genprimopcode_C_SRCS"))
* .build/stage0/utils/genprimopcode/package-data.mk
* .build/stage0/utils/genprimopcode/package-data.mk utils/genprimopcode/stage0/package-data.mk
Build system error - cannot currently call askOracle:
  Reason:          Within withResource using Resource ghc-cabal
  Question type:   OracleQ CmdLineFlags
  Question value:  OracleQ (CmdLineFlags ())
Move the askOracle call earlier/later
```

could you lend me a hand here?

**Comment by ndmitchell** [comment]
If you put the render box style in an oracle, then if someone decides to switch from `dot` to `unicorn` then suddenly its going to rebuild everything just so you can see lots of unicorns. I think the issue here is that this flag has gone in the oracle at all.

**Comment by snowleopard** [comment]
Looks nice! Although I expected unicorns to be different each time... :D 
Once the oracle issue is resolved we'll merge this.

**Comment by snowleopard** [comment]
@ndmitchell Agreed, a change in the output style shouldn't lead to rebuilds.

**Comment by ndmitchell** [comment]
Best suggestion is just to shove the render box style in a global IORef with the typical `unsafePerformIO $ newIORef ...` code. With HEAD Shake you could shove it in shakeExtra, but I'd still opt for an IORef here.

**Comment by angerman** [comment]
> Looks nice! Although I expected unicorns to be different each time... :D 

That's up to @bgamari :p I was only provided with one.

> @ndmitchell Agreed, a change in the output style shouldn't lead to rebuilds.

Right!

> Best suggestion is just to shove the render box style in a global IORef with the typical unsafePerformIO $ newIORef ... code. With HEAD Shake you could shove it in shakeExtra, but I'd still opt for an IORef here.

Wha! And I worked so hard to get the oracles right ;-)

File: src/Oracles/Config/CmdLineFlag.hs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    deriving (Show, Typeable, Eq, Hashable, Binary, NFData)
+
+buildInfo :: Action BuildInfoFlag
+buildInfo = do
+    opts <- askOracle $ CmdLineFlags ()
+    return $ flagBuildInfo opts
+
+cmdLineOracle :: [CmdLineOptions -> CmdLineOptions] -> Rules ()
+cmdLineOracle opts = do
+    cache <- newCache $ \_ -> return $ foldl (flip id) defaultCmdLineOptions opts
```
Comment (ndmitchell) [inline src/Oracles/Config/CmdLineFlag.hs]: I don't think the cache helps here since the oracle itself is cached

**Comment by ndmitchell** [comment]
The Oracle stuff looks good - I reviewed it and made a single minor remark for next time you have to go near oracles :)

**Comment by angerman** [comment]
> The Oracle stuff looks good - I reviewed it and made a single minor remark for next time you have to go near oracles :)

Thanks. Much appreciated!

**Comment by angerman** [comment]
Alright, now using an `IORef`. This feels dirty.  Once we add more command line options, we might want to distinguish between those that influence the build process and should end up in an Oracle, and those that should live in the `IORef`.

**Comment by snowleopard** [comment]
@angerman Thanks, I'll merge this.

I don't like using `unsafePerformIO` so I hope we'll figure out a nicer solution eventually.

**Pull request merged**